superduperdb/base/document.py

```diff
@@ -283,102 +285,135 @@ def _deep_flat_encode(
+        object = _deep_flat_decode(
+            builds[r['_object'][1:]], builds, getters=getters, db=db
+        )
         instance = _import_item(object=object.unpack(), dict=dict_, db=db)
         return instance
     if isinstance(r, dict):
+        literals = r.get('_literals', [])
         return {
-            k: _deep_flat_decode(v, cache, blobs, files, db=db) for k, v in r.items()
+            k: _deep_flat_decode(v, builds, getters=getters, db=db)
+            if k not in literals
+            else v
+            for k, v in r.items()
         }
-    if isinstance(r, str) and r.startswith('?') and not r.startswith('?db'):
-        return _get_leaf_from_cache(r, cache, blobs, files, db=db)
-    if isinstance(r, str) and r.startswith('&:'):
-        return _get_from_in_db_reference(r, cache, blobs, files, db=db)
-    if isinstance(r, str) and re.match("^\?db\.load\((.*)\)$", r):
-        match = re.match("^\?db\.load\((.*)\)$", r)
-        assert match is not None
-        assert db is not None, 'db is required for ?db.load()'
-        args = [x.strip() for x in match.groups()[0].split(',')]
-        if len(args) == 1:
-            return db.load(uuid=args[0])
-        if len(args) == 2:
-            return db.load(type_id=args[0], identifier=args[1])
-        if len(args) == 3:
-            return db.load(type_id=args[0], identifier=args[1], version=int(args[2]))
-        raise ValueError(f'Invalid number of arguments for {r}')
+    if isinstance(r, str) and r.startswith('?'):
+        return _get_leaf_from_cache(r[1:], builds, getters=getters, db=db)
+    if isinstance(r, str) and r.startswith('&'):
+        assert getters is not None
+        reference = parse_reference(r)
+        if reference.name in getters:
+            return getters[reference.name](reference.path)
+        return r
+    if isinstance(r, str) and (vars := re.findall(r'^<var:(.*?)>$', r)):
+        return Variable(vars[0])
+
     return r
 
 
-def _get_from_in_db_reference(
-    r, cache, blobs, files={}, db: t.Optional['Datalayer'] = None
-):
-    if not (reference := parse_reference(r)):
-        return r
+def _get_component(db, path):
+    parts = path.split(':')
+    if len(parts) == 1:
+        return db.load(uuid=parts[0])
+    if len(parts) == 2:
+        return db.load(type_id=parts[0], identifier=parts[1])
+    if len(parts) == 3:
+        if not parts[2].isnumeric():
+            return db.load(uuid=parts[2])
+        return db.load(type_id=parts[0], identifier=parts[1], version=parts[2])
+    raise ValueError(f'Invalid component reference: {path}')
 
-    if reference.name == 'component':
-        assert db is not None, 'db is required for ?:component:'
-        return db.load(uuid=r.split('/')[-1])
 
-    return r
+# TODO: How about the lazy loading of the artifact?
+# Should this function return a callback function for getting the artifact?
+def _get_artifact(db, path):
+    return db.artifact_store.get_bytes(path)
```

Comment:
@blythed The current logic for fetching bytes from the database is handled in the artifact, which parses the reference and then retrieves the bytes. Please confirm.

Reply:
Yes I like the callback approach. We could use this approach also for doing downloadable artifacts at some point.

Reply:
So you would "forward" a getter to the `Artifact`

Reply:
Done